File: go-appservice/internal/keyvault/keyvault.go
// Package keyvault provides pluggable encryption for secrets at rest.
// Cross-language compatible with Python agent/security/key_vault.py.
//
// Byte format:  prefix(1) || nonce(12) || ciphertext || tag(16)
// Prefix 0x01 = AES-256-GCM, 0x02 = HPKE-MLKEM (reserved).
package keyvault

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
)

const (
	prefixAESGCM byte = 0x01
	prefixHPKE   byte = 0x02 // reserved for HPKE-MLKEM backend
	nonceSize         = 12
)

// KeyVault encrypts and decrypts secrets at rest.
type KeyVault interface {
	Encrypt(plaintext string) ([]byte, error)
	Decrypt(ciphertext []byte) (string, error)
	Backend() string
}

// AESGCMVault implements AES-256-GCM encryption.
// Compatible with Python AESGCMVault (identical byte format).
type AESGCMVault struct {
	gcm cipher.AEAD
}

// NewAESGCMVault creates a vault from a 32-byte key (64 hex chars).
func NewAESGCMVault(secretHex string) (*AESGCMVault, error) {
	key, err := hex.DecodeString(secretHex)
	if err != nil {
		return nil, fmt.Errorf("keyvault: invalid hex secret: %w", err)
	}
	if len(key) != 32 {
		return nil, fmt.Errorf("keyvault: secret must be 32 bytes (64 hex chars), got %d", len(key))
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("keyvault: aes cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("keyvault: gcm: %w", err)
	}

	return &AESGCMVault{gcm: gcm}, nil
}

func (v *AESGCMVault) Encrypt(plaintext string) ([]byte, error) {
	nonce := make([]byte, nonceSize)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, fmt.Errorf("keyvault: nonce: %w", err)
	}

	ct := v.gcm.Seal(nil, nonce, []byte(plaintext), nil)

	// prefix(1) || nonce(12) || ciphertext+tag
	out := make([]byte, 0, 1+nonceSize+len(ct))
	out = append(out, prefixAESGCM)
	out = append(out, nonce...)
	out = append(out, ct...)
	return out, nil
}

func (v *AESGCMVault) Decrypt(ciphertext []byte) (string, error) {
	if len(ciphertext) < 1+nonceSize+v.gcm.Overhead() {
		return "", errors.New("keyvault: ciphertext too short")
	}

	prefix := ciphertext[0]
	if prefix == prefixHPKE {
		return "", errors.New("keyvault: HPKE-MLKEM ciphertext — use HPKEVault to decrypt")
	}
	if prefix != prefixAESGCM {
		return "", fmt.Errorf("keyvault: unknown prefix 0x%02x", prefix)
	}

	nonce := ciphertext[1 : 1+nonceSize]
	ct := ciphertext[1+nonceSize:]

	plain, err := v.gcm.Open(nil, nonce, ct, nil)
	if err != nil {
		return "", fmt.Errorf("keyvault: decrypt: %w", err)
	}

	return string(plain), nil
}

func (v *AESGCMVault) Backend() string { return "aesgcm" }

// NewKeyVault creates the appropriate vault based on backend name.
func NewKeyVault(backend, secretHex string) (KeyVault, error) {
	switch backend {
	case "aesgcm", "":
		return NewAESGCMVault(secretHex)
	case "hpke-mlkem":
		return NewHPKEVault(secretHex)
	default:
		return nil, fmt.Errorf("keyvault: unknown backend %q (supported: aesgcm, hpke-mlkem)", backend)
	}
}

File: python-backend/agent/security/key_vault.py
"""KeyVault — AES-256-GCM encryption for API keys at rest.

Cross-language compatible with Go internal/keyvault/ (identical byte format).
Format: 0x01 || 12-byte nonce || ciphertext || 16-byte GCM tag

ENV: KEY_ENCRYPTION_SECRET (64 hex chars = 32 bytes)
     KEY_VAULT_BACKEND (default: aesgcm)
"""

from __future__ import annotations

import os
from typing import Protocol, runtime_checkable

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

# Prefix bytes for auto-detection during decrypt
_PREFIX_AESGCM = b"\x01"
_PREFIX_HPKE = b"\x02"  # reserved for future PQC backend

_NONCE_SIZE = 12  # AES-GCM standard nonce size


@runtime_checkable
class KeyVault(Protocol):
    """Pluggable encryption interface for secrets at rest."""

    def encrypt(self, plaintext: str) -> bytes: ...
    def decrypt(self, ciphertext: bytes) -> str: ...
    @property
    def backend(self) -> str: ...


class AESGCMVault:
    """AES-256-GCM encryption. Cross-language compatible with Go keyvault."""

    def __init__(self, secret_hex: str) -> None:
        if len(secret_hex) != 64:
            raise ValueError(
                "KEY_ENCRYPTION_SECRET must be 64 hex chars (32 bytes). "
                'Generate with: python -c "import secrets; print(secrets.token_hex(32))"'
            )
        self._key = bytes.fromhex(secret_hex)
        self._aesgcm = AESGCM(self._key)

    def encrypt(self, plaintext: str) -> bytes:
        nonce = os.urandom(_NONCE_SIZE)
        ct = self._aesgcm.encrypt(nonce, plaintext.encode("utf-8"), None)
        return _PREFIX_AESGCM + nonce + ct

    def decrypt(self, ciphertext: bytes) -> str:
        if not ciphertext:
            raise ValueError("Empty ciphertext")
        prefix = ciphertext[0:1]
        if prefix == _PREFIX_HPKE:
            raise NotImplementedError(
                "HPKE-MLKEM decrypt not yet implemented in Python. Use Go backend."
            )
        if prefix != _PREFIX_AESGCM:
            raise ValueError(f"Unknown encryption prefix: {prefix!r}")
        nonce = ciphertext[1 : 1 + _NONCE_SIZE]
        ct = ciphertext[1 + _NONCE_SIZE :]
        return self._aesgcm.decrypt(nonce, ct, None).decode("utf-8")

    @property
    def backend(self) -> str:
        return "aesgcm"


# Singleton
_vault: KeyVault | None = None


def get_vault() -> KeyVault:
    """Get or create the global KeyVault singleton from ENV."""
    global _vault
    if _vault is not None:
        return _vault

    secret = os.environ.get("KEY_ENCRYPTION_SECRET", "")
    if not secret:
        raise RuntimeError(
            "KEY_ENCRYPTION_SECRET not set. API key encryption requires this ENV var. "
            'Generate with: python -c "import secrets; print(secrets.token_hex(32))"'
        )

    backend = os.environ.get("KEY_VAULT_BACKEND", "aesgcm")
    if backend == "aesgcm":
        _vault = AESGCMVault(secret)
    else:
        raise ValueError(f"Unknown KEY_VAULT_BACKEND: {backend}. Supported: aesgcm")

    return _vault
